splitgraph/core/fragment_manager.py

```diff
@@ -641,6 +674,138 @@ def record_table_as_patch(
+            )
+            + object_select
+            + SQL(")")
+        )
+
+        # Calculate the object ID, continue if there are deleted or inserted rows
+        schema_hash = self._calculate_schema_hash(new_schema_spec)
+
+        deletion_hash, rows_deleted = self.calculate_content_hash(
+            "pg_temp",
+            tmp_table,
+            new_schema_spec,
+            chunk_condition_sql=SQL("WHERE {} IS FALSE").format(Identifier(SG_UD_FLAG)),
+        )
+        insertion_hash, rows_inserted = self.calculate_content_hash(
+            "pg_temp",
+            tmp_table,
+            new_schema_spec,
+            chunk_condition_sql=SQL("WHERE {} IS TRUE").format(Identifier(SG_UD_FLAG)),
+        )
+
+        if rows_inserted != 0 or rows_deleted != 0:
+            content_hash = (insertion_hash - deletion_hash).hex()
+
+            object_id = "o" + sha256((content_hash + schema_hash).encode("ascii")).hexdigest()[:-2]
+
+            self._store_object_from_temp_table(
+                object_id,
+                tmp_table,
+                old_table,
+                in_fragment_order=in_fragment_order,
+                overwrite=overwrite,
+            )
+
+            with self.metadata_engine.savepoint("object_register"):
+                try:
+                    # Register the object: this also indexes it and (in the case of our custom object
+                    # manager) marks it as external/uploads it to S3.
+                    self._register_object(
+                        object_id,
+                        namespace=old_table.repository.namespace,
+                        insertion_hash=insertion_hash.hex(),
+                        deletion_hash=deletion_hash.hex(),
+                        table_schema=new_schema_spec,
+                        extra_indexes=extra_indexes,
+                        rows_inserted=rows_inserted,
+                        rows_deleted=rows_deleted,
+                    )
+                except UniqueViolation:
+                    logging.info(
+                        "Object %s already exists, continuing...",
+                        object_id,
+                    )
+
+            self.register_tables(
+                old_table.repository,
+                [
+                    (
+                        image_hash,
+                        old_table.table_name,
+                        new_schema_spec,
+                        old_table.objects + [object_id],
+                    )
+                ],
+            )
+
+        # Truncate the upper table to prepare for the next write
+        self.object_engine.run_sql(
+            SQL("TRUNCATE TABLE {}.{}").format(Identifier(schema), Identifier(upper_table))
+        )
+
+        # Finally, point the LQ FDW to the latest image to pick up the new changes after truncating the upper table
+        self.object_engine.run_sql(
+            SQL("ALTER SERVER {} OPTIONS (SET image_hash %s)").format(
```

Comment:
Does this get picked up by the LQFDW? I would have thought that the image hash is stored inside of the LQFDW class instance, which is initialized only on first query to the foreign table on a new connection.

Reply:
Well during development I found that this step was required in order to prevent the lower table (and hence the overlay view) from showing the previous content even after committing new/pending changes, so it seems to get picked up.

Reply:
So I see now that LQFDW gets re-initialized after commit (i.e. after altering server options), which explains why it works. 

It could be that PG is smart enough to realize that changing these options warrants a re-initialization.

Reply:
Correction, Multicorn is smart enough to [re-initialize](https://github.com/splitgraph/Multicorn/blob/master/src/python.c#L591) the FDW instance on options changes.

Reply:
Nice! looks like this is all fine then